### ISSUE TITLE
struct simple causes duplicate lines in experiment.xml

### ISSUE THREAD

**Issue by mwkrentel** [HPCToolkit/hpctoolkit#145]
There appears to be a bad interaction between hpcprof and Struct
Simple that causes duplicate lines in the experiment.xml file.

Here is an example with hpcstruct-bin analyzing itself on x86_64
(mic.cs).  This is probably overkill, probably you can make a much
simpler unit test.  (And that would be a good first step in
investigating the problem.)

Anyway, running hpcprof with a .hpcstruct file for hpcstruct-bin gives
this in the viewer.

![image1](https://user-images.githubusercontent.com/3969036/63665307-64b03b80-c790-11e9-9fc0-f079b013e8cb.png)

And running hpcprof on the same measurements directory and NO struct
file gives duplicate entries for BAnal::Struct::makeStructure().

![image2](https://user-images.githubusercontent.com/3969036/63665323-7691de80-c790-11e9-8d8d-7aa2b620c419.png)

The problem is not the viewer.  You can grep the experiment.xml file
for 'makeStructure' and there are duplicates in the second example but
not the first.

I don't know if the problem is in hpcprof or in Struct Simple or in
some bad interaction between the two.

This is probably the real cause of (1) in issue #143.

A wild guess.  Note that all but the first makeStructure() lines in
the second example have just a handful of samples.  Maybe hpcrun is
generating some one-off PC addresses that the .hpcstruct file
understands but are missed by Struct Simple.  That could explain why
prof thinks they're from a different function (with the same name).
But that's just a wild guess.

**Comment by mwkrentel** [comment]
@blue42u (Jonathon Anderson) reported that this problem exists with
binutils 2.31.1 but not with 2.32.  I tried a few more revs.

```
binutils-2.27     works
binutils-2.29.1   works
binutils-2.31.1   duplicate functions
binutils-2.32     works
```

I haven't tried 2.30 or 2.31 or their git repo, but it looks like this
was a short-lived bug that just happened to affect the version that we
were using.

So, we'll do a smoke test with 2.32 to make sure it works on all
platforms, etc and update the packages.yaml file and add a note to
avoid 2.31.1.  We'll need 2.32 at some point anyway as Linux distros
start to use it (eg, the Arch Linux issue).

----------

This much I can believe.  Someone added a bug and it got fixed.  But
the idea that this was a 30 year old bug that just got fixed now was
just too incredible.

I'm a bit curious for more details about the mechanics of the bug.
But probably not so much that I'll go bisecting the binutils repo.

And I wouldn't rule out other or related bugs in our code.  Bill
Williams has told me that you often have to interpret the raw binutils
data with offsets and starting and ending ranges, etc.  This is the
kind of thing that Symtab does but I don't know if our code does.

**Comment by blue42u** [comment]
A few more revs:
```
binutils-2.29.1  works
binutils-2.30    duplicates
binutils-2.31    duplicates
binutils-2.31.1  duplicates
binutils-2.32    works
```
which means it went unnoticed for around 1.5yr. Which leads me to believe its an interpretation issue.

Moving forward, I think the best approach will be to convert the `libbfd` dependent code to use `libdw` or Symtab instead. We already require both as project dependencies, for hpcprof it shouldn't matter as much which is used (Symtab may reduce the number of heuristics needed, but `libdw` is simpler and pure C).

**Comment by mwkrentel** [comment]
1. This is fixed with binutils 2.32.
2. Binutils 2.32 is now in spack.
3. I've updated the toolkit packages.yaml file to binutils 2.32.

So, I'm going to declare this as resolved and close.
